metrics_utility/automation_controller_billing/collectors.py

```diff
@@ -563,42 +571,89 @@ def total_workers_vcpu(since, full_path, until, **kwargs):
+        raise MetricsException(f'Can not create a prometheus api client ERROR: {e}')
 
     try:
-        nodes = api_instance.list_node()
-    except Exception as e:
+        total_workers_vcpu, promql_query = get_total_workers_cpu(prom, prev_hour_start)
+        timeline = get_cpu_timeline(prom, prev_hour_start, prev_hour_end)
+    except MetricsException as e:
         raise MetricsException(f'Unexpected error when retrieving nodes: {e}')
 
-    if nodes is None:
-        raise MetricsException('No nodes found')
+    info['promql_query'] = promql_query
+    info['timeline'] = timeline
+
+    logger.debug(f'total_workers_vcpu: {total_workers_vcpu}')
 
-    total_workers_vcpu = 0
-    # In SaaS case we have only Worker nodes and so we don't need to filter out the control plan.
-    # If it used for other environement, we might need to implement the filtering.
-    for node_info in nodes.items:
-        for resource, value in node_info.status.capacity.items():
-            if resource == 'cpu':
-                info['nodes'].append({node_info.metadata.name: int(value)})
-                total_workers_vcpu += int(value)
+    # This can happen when the prev_hour_start doesn't have data, it could be when the cluster just started or
+    # if for some reasons prometheus loss some data.
+    if total_workers_vcpu is None:
+        logger.warning('No data availble yet, the cluster is probably running for less than an hour')
+        raise MetricsException('No data availble yet, the cluster is probably running for less than an hour')
 
-    info['total_workers_vcpu'] = total_workers_vcpu
+    info['total_workers_vcpu'] = int(total_workers_vcpu)
 
     # This message must always appear in the log regardless of the log level.
     logger_info_level.info(json.dumps(info, indent=2))
 
-    return {'cluster_name': info['cluster_name'], 'total_workers_vcpu': info['total_workers_vcpu']}
+    return {'timestamp': info['timestamp'], 'cluster_name': info['cluster_name'], 'total_workers_vcpu': info['total_workers_vcpu']}
+
+
+def get_hour_boundaries(current_timestamp: float) -> Tuple[float, float, float]:
+    current_hour_start = (current_timestamp // 3600) * 3600
+    previous_hour_start = current_hour_start - 3600
+    previous_hour_end = current_hour_start - 1
+    return previous_hour_start, previous_hour_end, current_hour_start
+
+
+def get_total_workers_cpu(prom: PrometheusClient, base_timestamp: float) -> Tuple[float, str]:
+    promql_query = f'max_over_time(sum(machine_cpu_cores)[59m59s:5m] @ {base_timestamp})'
+
+    try:
+        total_workers_vcpu = prom.get_current_value(promql_query)
+    except Exception as e:
+        raise MetricsException(f'Unexpected error when retrieving nodes: {e}')
+
+    return total_workers_vcpu, promql_query
+
+
+def get_cpu_timeline(prom: PrometheusClient, previous_hour_start, previous_hour_end: float) -> list:
+    """
+    Get array of timestamp/CPU pairs for the hour leading up to base_timestamp
```

Comment:
Nit: base_timestamp isn't a param for this function. Please update to reflect current behavior.

Reply:
Done